src/main/scala/is/hail/expr/ir/functions/Functions.scala

```diff
@@ -0,0 +1,180 @@
+package is.hail.expr.ir.functions
+
+import is.hail.asm4s._
+import is.hail.expr.{AST, FunType, Lambda, TypeTag}
+import is.hail.expr.ir._
+import is.hail.expr.types._
+import is.hail.utils._
+
+import scala.collection.mutable
+import scala.reflect.ClassTag
+
+object IRFunctionRegistry {
+
+  val registry: mutable.Map[String, Seq[(TypeTag, Seq[IR] => IR)]] = mutable.Map().withDefaultValue(Seq.empty)
+
+  UtilFunctions.registerAll(registry)
+
+  def lookupFunction(name: String, args: TypeTag): Option[Seq[IR] => IR] = {
+    val validMethods = registry(name).flatMap { case (tt, f) =>
+      if (tt.xs.length == args.xs.length) {
+        tt.clear()
+        if (tt.unify(args))
+          Some(f)
+        else
+          None
+      }
+      else {
+        None
+      }
+    }
+    validMethods match {
+      case Seq() => None
+      case Seq(x) => Some(x)
+      case _ => fatal(s"Multiple methods found that satisfy $name$args.")
+    }
+  }
+}
+
+abstract class RegistryFunctions {
+
+  def registerAll(r: mutable.Map[String, Seq[(TypeTag, Seq[IR] => IR)]]) {
+    registry.foreach { case (k, v) =>
+      r.put(k, r(k) ++ registry(k))
+    }
+  }
+
+  private val registry = mutable.Map[String, Seq[(TypeTag, Seq[IR] => IR)]]().withDefaultValue(Seq.empty)
+
+  private var tvs = mutable.Map[String, TVariable]()
+
+  def tv(name: String): TVariable = {
+    tvs.getOrElse(name, {
+      tvs.put(name, TVariable(name))
+      tvs(name)
+    })
+  }
+
+  def addIRFunction(f: IRFunction) {
+    val l = registry(f.name)
+    registry.put(f.name,
+      l :+ (f.types, { args: Seq[IR] =>
+        ApplyFunction(f, args)
+      }))
+  }
+
+  def addIR(name: String, types: Seq[Type], f: Seq[IR] => IR) {
+    val l = registry(name)
+    registry.put(name, l :+ ((FunType(types: _*), f)))
+  }
+
+  def registerCode[R](mname: String, mtypes: Array[Type])(impl: (MethodBuilder, Array[Code[_]]) => Code[R]) {
+    addIRFunction(new IRFunction {
+      override val name: String = mname
+
+      override val types: TypeTag = FunType(mtypes.init: _*)
+
+      override val returnType: Type = mtypes.last
+
+      override def apply(mb: MethodBuilder, args: Code[_]*): Code[R] = impl(mb, args.toArray)
+    })
+  }
+
+  def registerScalaFunction[R: ClassTag](mname: String, types: Type*)(obj: Any, method: String)
+  {
+    registerCode[R](mname, types.toArray)
+      { (mb, args) =>
+        val cts = types.init.map(TypeToIRIntermediateClassTag(_).runtimeClass).toArray
+        Code.invokeScalaObject[R](obj, method, cts, args)
+      }
+  }
+
+  def registerJavaStaticFunction[C: ClassTag, R: ClassTag](mname: String, types: Type*)(method: String) {
+    registerCode[R](mname, types.toArray)
+      { (mb, args) =>
+        val cts = types.init.map(TypeToIRIntermediateClassTag(_).runtimeClass)
+        Code.invokeStatic[C, R](method, cts.toArray, args)
+      }
+  }
+
+  def registerIR(mname: String, types: Array[Type])(f: Seq[IR] => IR) {
+    addIR(mname, types.init, f)
```

Comment:
It seems cleaner to not put the return type in `types`, you end up pulling them apart all over the place.  It also isn't clear from a code readability perspective.  Which is cleaner, `types.init` or `argumentTypes`?  Or `types.tail` or `returnType`?